### PR TITLE
Unescape interpolated string literal components

### PR DESCRIPTION
**Pull request by 333fred**
Fixes https://github.com/dotnet/roslyn/issues/54703. Spec change is at https://github.com/dotnet/csharplang/pull/4910.

Note that there is an interesting wrinkle here in that, for an interpolated string used as a string, the change in constant value makes the lowering mechanism an observable side effect. While doing this step during lowering instead of initial binding is technically feasible, it would be quite complex as there are several different ways an AppendLiteral call could get to lowering (dynamic, conversion to object, regular literal, passed by in, etc). For now, I'm opting for simple strategy of accepting this observable side-effect, and if we want to try and make that side effect invisible we can look at doing so at a later point.

Test plan https://github.com/dotnet/roslyn/issues/51499

### CODE REVIEW

**Review by jcouv — APPROVED** [review]
LGTM Thanks (iteration 1)

File: src/Compilers/CSharp/Test/Semantic/Semantics/InterpolationTests.cs [hunk 1]
```diff
@@ -11019,5 +11019,101 @@ public static class MemoryExtensions
...
+            verifier.VerifyIL("<top-level-statements-entry-point>", @"
+{
+  // Code size       56 (0x38)
+  .maxstack  3
+  .locals init (int V_0, //i
+                System.Runtime.CompilerServices.DefaultInterpolatedStringHandler V_1)
+  IL_0000:  ldc.i4.1
+  IL_0001:  stloc.0
+  IL_0002:  ldloca.s   V_1
+  IL_0004:  ldc.i4.6
```
Comment (jaredpar) [inline src/Compilers/CSharp/Test/Semantic/Semantics/InterpolationTests.cs:11046]: Should this still be 6? Given that the handler only ever sees the single brace shouldn't we be counting the `{{` and `}}` as 1 when calculating the length here?

**Review by jaredpar — APPROVED** [review]
Minor feedback on the length being passed to handler ctor. Don't think that should be blocking though.

Thanks for getting this done!

**Comment by 333fred** [comment]
@jaredpar @jcouv re-requesting review because the refactor for the correct length is slightly longer than I'm comfortable with just assuming you still sign off on.

File: src/Compilers/CSharp/Portable/Binder/Binder_InterpolatedString.cs [hunk 2]
```diff
@@ -230,24 +228,19 @@ bool tryBindAsHandlerType([NotNullWhen(true)] out BoundInterpolatedString? resul
...
             Debug.Assert(appendCalls.All(a => a is { HasErrors: true } or BoundCall { Arguments: { Length: > 0 } } or BoundDynamicInvocation));
-            int baseStringLength = 0;
-            int numFormatHoles = 0;
 
-            foreach (var currentPart in unconvertedInterpolatedString.Parts)
+            if (needToCheckConversionToObject)
             {
-                if (currentPart is BoundStringInsert insert)
+                TypeSymbol objectType = GetSpecialType(SpecialType.System_Object, diagnostics, unconvertedInterpolatedString.Syntax);
+                BindingDiagnosticBag conversionDiagnostics = BindingDiagnosticBag.GetInstance(withDiagnostics: true, withDependencies: false); ;
```
Comment (jaredpar) [inline src/Compilers/CSharp/Portable/Binder/Binder_InterpolatedString.cs:239]:
```suggestion
                BindingDiagnosticBag conversionDiagnostics = BindingDiagnosticBag.GetInstance(withDiagnostics: true, withDependencies: false);
```

**Review by jaredpar — APPROVED** [review]

File: src/Compilers/CSharp/Portable/Binder/Binder_InterpolatedString.cs (hunk 2 above)
Comment (jaredpar) [reply src/Compilers/CSharp/Portable/Binder/Binder_InterpolatedString.cs:239]: Only saw this cause I made the **exact** same mistake today and Chris caught me 

https://github.com/dotnet/roslyn/pull/54675#discussion_r666587369

**Review by stephentoub — COMMENTED** [review]
Thank you!

File: src/Compilers/CSharp/Test/Semantic/Semantics/InterpolationTests.cs (hunk 1 above)
Comment (jcouv) [reply src/Compilers/CSharp/Test/Semantic/Semantics/InterpolationTests.cs]: Thanks for catching this, Jared.
Fred, consider adding a more explicit validation for this (having the constructor log the value).

**Review by jcouv — APPROVED** [review]
LGTM Thanks (iteration 4)

**Pull request merged**